src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionTemplatesPass.php

```diff
@@ -35,44 +36,78 @@ class ResolveDefinitionTemplatesPass implements CompilerPassInterface
      */
     public function process(ContainerBuilder $container)
     {
-        $this->container = $container;
         $this->compiler = $container->getCompiler();
         $this->formatter = $this->compiler->getLoggingFormatter();
 
         foreach ($container->getDefinitions() as $id => $definition) {
+            $this->currentId = $id;
+
             // yes, we are specifically fetching the definition from the
             // container to ensure we are not operating on stale data
             $definition = $container->getDefinition($id);
-            if (!$definition instanceof DefinitionDecorator || $definition->isAbstract()) {
-                continue;
+            if ($definition instanceof DefinitionDecorator) {
+                $definition = $this->resolveDefinition($container, $definition);
+                $container->setDefinition($id, $definition);
             }
 
-            $this->resolveDefinition($id, $definition);
+            $definition->setArguments($this->resolveArguments($container, $definition->getArguments()));
+            $definition->setMethodCalls($this->resolveArguments($container, $definition->getMethodCalls()));
+            $definition->setProperties($this->resolveArguments($container, $definition->getProperties()));
         }
     }
 
     /**
+     * Resolves definition decorator arguments.
+     *
+     * @param ContainerBuilder $container The ContainerBuilder
+     * @param array            $arguments An array of arguments
+     *
+     * @return array
+     */
+    private function resolveArguments(ContainerBuilder $container, array $arguments)
+    {
+        foreach ($arguments as $k => $argument) {
+            if (is_array($argument)) {
+                $arguments[$k] = $this->resolveArguments($container, $argument);
+            } elseif ($argument instanceof Definition) {
+                if ($argument instanceof DefinitionDecorator) {
+                    $arguments[$k] = $argument = $this->resolveDefinition($container, $argument);
+                }
+                $argument->setArguments($this->resolveArguments($container, $argument->getArguments()));
+                $argument->setMethodCalls($this->resolveArguments($container, $argument->getMethodCalls()));
+                $argument->setProperties($this->resolveArguments($container, $argument->getProperties()));
```

Comment:
well, there is only a small difference, which is what you are doing with the resolved definition in case of DefinitionDecorator. This could be kept in both locations.
Adding support for anonymous decorators in more places would increase the duplication even more